// @communities-webruntime/design/src/modules/webruntimedesign/designmode/handlers/select-proxy-handlers.js
/**
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { isElementVisible } from '../shared/utils';
import {
    clearSectionProxyChangesIfNeeded,
    isSection,
    updateProxyForSection,
} from '../shared/section-utils';
import { configureLabel } from '../shared/proxy-utils';
import {
    BASE_Z_INDEX,
    COMPONENT_LABEL_ATTRIBUTE_NAME,
    ELEMENT_ID_ATTRIBUTE_NAME,
    PROXY_LABEL_CLASS,
} from '../shared/constants';

const PROXY_TOOLBOX_CLASS = 'proxy-select-toolbox';
const PROXY_TOOLBOX_ITEM_CLASS = 'proxy-select-toolbox-item';
const PROXY_TOOLBOX_DELETE_CLASS = 'proxy-select-toolbox-delete';

function selectProxyInitHandler({ element, postMessageManager }) {
    const label = document.createElement('div');
    label.classList.add(PROXY_LABEL_CLASS);
    Object.assign(label.style, {
        position: 'absolute',
        padding: '0 10px',
        'background-color': '#0059A7',
        color: 'white',
        height: '26px',
        'user-select': 'none',
        'border-radius': '3px 3px 0 0',
        display: 'flex',
        'align-items': 'center',
        'justify-content': 'center',
    });
    element.appendChild(label);

    // create a re-usable component proxy toolbox
    const toolbox = document.createElement('div');
    toolbox.classList.add(PROXY_TOOLBOX_CLASS);
    Object.assign(toolbox.style, {
        position: 'absolute',
        top: '0',
        right: '0',
        display: 'flex',
        'user-select': 'none',
    });
    element.appendChild(toolbox);

    // toolbox - delete
    const toolboxDelete = document.createElement('div');
    toolboxDelete.classList.add(PROXY_TOOLBOX_ITEM_CLASS, PROXY_TOOLBOX_DELETE_CLASS);
    toolboxDelete.textContent = 'x';
    Object.assign(toolboxDelete.style, {
        height: '20px',
        width: '20px',
        'background-color': '#0059A7',
        color: 'white',
        display: 'flex',
        'align-items': 'center',
        'justify-content': 'center',
        'line-height': '20px',
    });
    toolbox.appendChild(toolboxDelete);
    toolboxDelete.addEventListener('mouseup', () => {
        const itemId = element.getAttribute(ELEMENT_ID_ATTRIBUTE_NAME);
        postMessageManager.sendPostmessage(window.parent, 'delete-item', {
            itemid: itemId,
        });
    });

    return element;
}

function selectProxyEnableHandler({ element, params, postMessageManager }) {
    const component = params.payload.element;

    element.setAttribute(
        ELEMENT_ID_ATTRIBUTE_NAME,
        component.element.getAttribute(ELEMENT_ID_ATTRIBUTE_NAME)
    );

    const rect = component.element.getBoundingClientRect();
    const style = {
        display: '',
        width: `${rect.width}px`,
        height: `${rect.height}px`,
        'box-shadow': '0 0 0 2px #0059A7',
        position: 'absolute',
        top: `${window.scrollY + rect.top}px`,
        left: `${window.scrollX + rect.left}px`,
        zIndex: BASE_Z_INDEX,
        cursor: 'pointer',
    };
    Object.assign(element.style, style);

    // set data-item-id with key from component element
    element.setAttribute(ELEMENT_ID_ATTRIBUTE_NAME, component.key);

    if (isElementVisible(element) && isElementVisible(component.element)) {
        const label = element.querySelector(`.${PROXY_LABEL_CLASS}`);

        // for sections, we hide the label and add the click-to-add buttons
        if (isSection(component.element)) {
            updateProxyForSection(element, component.element, postMessageManager);
        } else {
            configureLabel(label, component.element.getAttribute(COMPONENT_LABEL_ATTRIBUTE_NAME));
            clearSectionProxyChangesIfNeeded(element);
        }
    }
}

function selectProxyDisableHandler({ element }) {
    element.style.display = 'none';
    clearSectionProxyChangesIfNeeded(element);
}

export { selectProxyDisableHandler, selectProxyEnableHandler, selectProxyInitHandler };
