// @communities-webruntime/services/dist/commonjs/template-generators/default-component-template-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@communities-webruntime/common");
const template_utils_1 = require("./template-utils");
class DefaultComponentTemplateGenerator {
    generate({ cmp, attributeSet, containerRegion, isRenderDesignMode, isThemeLayout, }) {
        const elem = common_1.moduleSpecifierToElementName(cmp.descriptor);
        const attrs = template_utils_1.attributes(cmp, attributeSet);
        // If we're in designmode, the component isn't rendered in a slot but instead is rendered
        // directly by the webruntime-generated-component wrapper
        const slot = !isRenderDesignMode ? template_utils_1.buildSlot(containerRegion) : ` `;
        const rgns = template_utils_1.regions({
            regionList: cmp.regions,
            isThemeLayout: !!isThemeLayout,
            attributeSet: attrs.attributes,
            isRenderDesignMode,
        });
        return {
            html: `<${elem}${attrs.html}${slot}>${rgns}</${elem}>`,
            attrs,
        };
    }
    // No componentName, as it represents all components without a custom generator
    getComponentName() {
        return ``;
    }
}
exports.DefaultComponentTemplateGenerator = DefaultComponentTemplateGenerator;
