// @webruntime/server/dist/commonjs/bootstrap/bootstrap-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function serialize(resourceMap, nonce) {
    const { resources } = resourceMap;
    return resources.map(resource => {
        return `\n${resourceToHtml(resource, nonce)}`;
    });
}
exports.serialize = serialize;
function resourceToHtml(resource, sourceNonce) {
    if (resource.type === 'text/css') {
        if (resource.code) {
            return getStyleTagCode(resource, sourceNonce);
        }
        return getStyleTagHref(resource);
    }
    if (resource.code) {
        if (resource.type.startsWith('webruntime')) {
            // no nonce needed for non-browser types
            return getScriptTagCode(resource);
        }
        return getScriptTagCode(resource, sourceNonce);
    }
    if (resource.useLinkPreload) {
        return getLinkPreload(resource);
    }
    return getScriptTagHref(resource);
}
function getStyleTagCode(style, sourceNonce) {
    return `<style type="${style.type}"${style.media ? ` media="${style.media}"` : ''}${sourceNonce ? ` nonce="${sourceNonce}"` : ''}>${style.code}</style>`;
}
function getStyleTagHref(style) {
    return `<link rel="stylesheet" type="${style.type}"${style.media ? ` media="${style.media}"` : ''} href="${style.path}">`;
}
function getScriptTagCode(script, sourceNonce) {
    return `<script type="${script.type}"${sourceNonce ? ` nonce="${sourceNonce}"` : ''}>${script.code}</script>`;
}
function getScriptTagHref(script) {
    const attributes = script.attributes ? script.attributes.join(' ') : null;
    return `<script type="${script.type}" ${attributes ? `${attributes} ` : ''}src="${script.path}"></script>`;
}
function getLinkPreload(resource) {
    return `<link rel="preload" crossorigin="anonymous" as="script" href="${resource.path}">`;
}
