// @communities-webruntime/extensions/dist/commonjs/file-loggers.js
"use strict";
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class Logger {
    constructor(filename) {
        this.filepath = path_1.default.resolve(`./${filename}`);
    }
    clear() {
        if (fs_1.default.existsSync(this.filepath)) {
            fs_1.default.unlinkSync(this.filepath);
        }
    }
    log(data) {
        fs_1.default.appendFileSync(this.filepath, `${JSON.stringify(data)}\n`);
    }
}
exports.Logger = Logger;
const METRICS_LOG_FILE = 'metrics.log';
const CLIENT_ERROR_LOG_FILE = 'clienterror.log';
const METRICS_LOGGER = new Logger(METRICS_LOG_FILE);
exports.METRICS_LOGGER = METRICS_LOGGER;
const CLIENT_ERROR_LOGGER = new Logger(CLIENT_ERROR_LOG_FILE);
exports.CLIENT_ERROR_LOGGER = CLIENT_ERROR_LOGGER;
