// @webruntime/api/dist/commonjs/messages.js
"use strict";
/** @hidden */
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@lwc/errors");
exports.configMessages = {
    INVALID_COMPILE_MODE: {
        code: 1000,
        message: 'Warning: "{0}" is not a valid compile mode, defaulting to "{1}"',
        level: errors_1.DiagnosticLevel.Warning,
        url: '',
    },
    MISSING_PROJECTDIR: {
        code: 1001,
        message: 'Cannot find projectDir at "{0}"',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_RESOURCE_ROOT: {
        code: 1002,
        message: 'If defined, server.resourceRoot must start with "/", but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_BASE_PATH: {
        code: 1003,
        message: 'If defined, server.basePath must start with "/", but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_MODULE_FORMAT: {
        code: 1004,
        message: 'Unsupported module format: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_MODULE_FORMAT_ESM: {
        code: 1005,
        message: 'ESM format is supported in "dev" mode only',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_CONFIG_FILES: {
        code: 1006,
        message: 'configFiles must be an array of strings, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_BUILD_DIR: {
        code: 1007,
        message: 'Warning: {0} is not a valid string buildDir, defaulting to "{1}"',
        level: errors_1.DiagnosticLevel.Warning,
        url: '',
    },
    INVALID_PORT: {
        code: 1008,
        message: 'Warning: The configured port value is not a number, defaulting to: {0}',
        level: errors_1.DiagnosticLevel.Warning,
        url: '',
    },
    INVALID_MODULE_DIR: {
        code: 1009,
        message: 'If defined, moduleDir must be a string, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_LOCALE: {
        code: 1010,
        message: 'Warning: {0} is not a valid defaultLocale, defaulting to "{1}"',
        level: errors_1.DiagnosticLevel.Warning,
        url: '',
    },
    INVALID_DEFAULT_CMP: {
        code: 1011,
        message: 'If defined, app.defaultComponent must be a string, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_DEFAULT_TEMPLATE: {
        code: 1012,
        message: 'If defined, app.defaultTemplate must be a string, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_PRELOAD_MODULES: {
        code: 1013,
        message: 'preloadModules must be an array of strings, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_EXTERNALS: {
        code: 1014,
        message: 'externals must be an array of strings, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_BUNDLE: {
        code: 1015,
        message: 'bundle must be an array of strings or { descriptor, exclude[] } objects, but is: {0}',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
};
