// @webruntime/api/dist/commonjs/webruntime-app.js
"use strict";
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The Base LWR Page Class which the default LWR page and custom pages extend.
 *
 * Create a custom Page class extended from WebruntimePage:
 * @example
 *```js
 * const {WebruntimeAppDefinition} = require('@webruntime/api');
 * const SamplePage = require('src/server/sample-page.js');
 *
 * class SampleApp extends WebruntimeAppDefinition {
 *  get pages() {
 *      return [{
 *          route: '/*',
 *          page: SamplePage
 *      }]
 *  }
 * }
 * ```
 */
class WebruntimeAppDefinition {
    // eslint-disable-next-line no-useless-constructor, no-empty-function
    constructor(config) {
        this.config = config;
    }
    get pages() {
        return [];
    }
    get defaultComponent() {
        return this.config.defaultComponent;
    }
    get defaultTemplate() {
        return this.config.defaultTemplate;
    }
}
exports.WebruntimeAppDefinition = WebruntimeAppDefinition;
