// @communities-webruntime/design/src/modules/webruntimedesign/designmode/handlers/drag-proxy-handlers.js
/**
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { BASE_Z_INDEX } from '../shared/constants';

function dragProxyEnableHandler({ element, params }) {
    const component = params.payload.element;
    const rect = component.element.getBoundingClientRect();

    const style = {
        display: 'block',
        position: 'fixed',
        left: `${params.payload.x - rect.width / 2}px`,
        top: `${params.payload.y - rect.height / 2}px`,
        height: `${rect.height}px`,
        width: `${rect.width}px`,
        border: '2px solid #d4d4d4',
        backgroundColor: '#f3f3f3',
        opacity: 0.5,
        zIndex: BASE_Z_INDEX + 3,
    };

    Object.assign(element.style, style);
}

function dragProxyDisableHandler({ element }) {
    if (element && element.style) {
        element.style.display = 'none';
    }
}

export { dragProxyDisableHandler, dragProxyEnableHandler };
