// @salesforce/lwc-dev-server-dependencies/vendors/dependencies-228/lightning-pkg/src/lightning/select/select.js
import labelRequired from '@salesforce/label/LightningControl.required';
import { api, LightningElement, track } from 'lwc';
import { classSet } from 'lightning/utils';
import {
    classListMutation,
    getRealDOMId,
    normalizeBoolean,
    synchronizeAttrs,
} from 'lightning/utilsPrivate';
import {
    FieldConstraintApi,
    InteractingState,
    isEmptyString,
    normalizeVariant,
    VARIANT,
} from 'lightning/inputUtils';

const i18n = {
    required: labelRequired,
};

export default class LightningSelect extends LightningElement {
    static delegatesFocus = true;

    @api label;
    @api name;
    @api messageWhenValueMissing;
    @api accessKey;
    @api autocomplete;

    @track _options = [];

    _helpMessage = '';
    _selectedValue;
    _variant;
    _required = false;
    _disabled = false;
    _multiple = false;
    _fieldLevelHelp;
    _size;

    connectedCallback() {
        this.classList.add('slds-form-element');
        this.updateClassList();
        this.interactingState = new InteractingState();
        this.interactingState.onleave(() => this.reportValidity());
    }

    renderedCallback() {
        if (this.options && this._selectedValue !== undefined) {
            this.selectOptionsByValue(this._selectedValue);
        }
        synchronizeAttrs(this.selectElement, {
            'aria-describedby': getRealDOMId(
                this.template.querySelector('#help-message')
            ),
        });
    }

    @api
    get fieldLevelHelp() {
        return this._fieldLevelHelp;
    }

    set fieldLevelHelp(value) {
        this._fieldLevelHelp = value;
    }

    @api
    get variant() {
        return this._variant || VARIANT.STANDARD;
    }

    set variant(value) {
        this._variant = normalizeVariant(value);
        this.updateClassList();
    }

    @api
    get multiple() {
        return this._multiple;
    }

    set multiple(value) {
        this._multiple = normalizeBoolean(value);
    }

    @api
    get size() {
        if (!this.multiple) {
            return null;
        }

        if (this._size === undefined) {
            return '4';
        }
        return this._size;
    }

    set size(newValue) {
        this._size = newValue;
    }

    @api
    get required() {
        return this._required;
    }

    set required(value) {
        this._required = normalizeBoolean(value);
    }

    @api
    get disabled() {
        return this._disabled;
    }

    set disabled(value) {
        this._disabled = normalizeBoolean(value);
    }

    @api
    get value() {
        return this._selectedValue;
    }

    set value(newValue) {
        this._selectedValue = newValue;

        if (this.isConnected && newValue) {
            this.selectOptionsByValue(newValue);
        }
    }

    @api
    get options() {
        return this._options;
    }

    set options(newValue) {
        this._options = newValue;

        if (this.isConnected && newValue) {
            this.selectOptionsByValue(this._selectedValue);
        }
    }

    updateClassList() {
        classListMutation(this.classList, {
            'slds-form-element_stacked': this.variant === VARIANT.LABEL_STACKED,
            'slds-form-element_horizontal':
                this.variant === VARIANT.LABEL_INLINE,
        });
    }

    @api
    focus() {
        if (this.isConnected) {
            this.selectElement.focus();
        }
    }

    @api
    blur() {
        if (this.isConnected) {
            this.selectElement.blur();
        }
    }

    /**
     * Represents the validity states that an element can be in, with respect to constraint validation.
     * @type {object}
     *
     */
    @api
    get validity() {
        return this._constraint.validity;
    }

    /**
     * Checks if the input is valid.
     * @returns {boolean} Indicates whether the element meets all constraint validations.
     */
    @api
    checkValidity() {
        return this._constraint.checkValidity();
    }

    /**
     * Sets a custom error message to be displayed when a form is submitted.
     * @param {string} message - The string that describes the error. If message is an empty string, the error message is reset.
     */
    @api
    setCustomValidity(message) {
        this._constraint.setCustomValidity(message);
    }

    /**
     * Displays the error messages and returns false if the input is invalid.
     * If the input is valid, reportValidity() clears displayed error messages and returns true.
     * @returns {boolean} - The validity status of the input fields.
     */
    @api
    reportValidity() {
        return this._constraint.reportValidity(message => {
            this._helpMessage = message;
        });
    }

    get i18n() {
        return i18n;
    }

    get selectElement() {
        return this.template.querySelector('select');
    }

    get isLabelHidden() {
        return this.variant === VARIANT.LABEL_HIDDEN;
    }

    get computedLabelClass() {
        return classSet('slds-form-element__label')
            .add({ 'slds-assistive-text': this.isLabelHidden })
            .toString();
    }

    get _constraint() {
        if (!this._constraintApi) {
            this._constraintApi = new FieldConstraintApi(() => this, {
                valueMissing: () => {
                    const isMultipleAndNoValue =
                        this.multiple &&
                        (!this.value || this.value.length === 0);
                    return (
                        !this.disabled &&
                        this.required &&
                        (isEmptyString(this.value) || isMultipleAndNoValue)
                    );
                },
            });
        }
        return this._constraintApi;
    }

    handleChange(event) {
        event.preventDefault();
        event.stopPropagation();

        this._selectedValue = this.getSelectedOptionValues();

        this.dispatchEvent(
            new CustomEvent('change', {
                bubbles: true,
                composed: true,
                detail: {
                    value: this._selectedValue,
                },
            })
        );
    }

    handleFocus() {
        this.interactingState.enter();

        this.dispatchEvent(new CustomEvent('focus'));
    }

    handleBlur() {
        this.interactingState.leave();

        this.dispatchEvent(new CustomEvent('blur'));
    }

    selectOptionsByValue(optionValue) {
        if (this.multiple) {
            if (Array.isArray(optionValue)) {
                const options = this.template.querySelectorAll('option');
                options.forEach(option => {
                    option.selected = optionValue.includes(option.value);
                });
            }
        } else {
            this.selectElement.value = optionValue;
        }
    }

    getSelectedOptionValues() {
        if (this.multiple) {
            const options = this.template.querySelectorAll('option');
            return Array.prototype.reduce.call(
                options,
                (selectedValues, option) => {
                    if (option.selected) {
                        selectedValues.push(option.value);
                    }
                    return selectedValues;
                },
                []
            );
        }
        return this.selectElement.value;
    }
}
