// @webruntime/navigation/dist/commonjs/services/messages.js
"use strict";
/** @hidden */
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@lwc/errors");
exports.messages = {
    EMPTY_ROUTE_DATA: {
        code: 4010,
        message: 'Route data validation failed: Route with id "{0}" from {1}.json must have a non-empty "{2}" property',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
    MISSING_ROUTE_PATH: {
        code: 4011,
        message: 'Route data file does not exist at path "{0}"',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    MISSING_ROUTE_SET: {
        code: 4012,
        message: 'Cannot find route set with id "{0}"',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
    MISSING_ROUTE_DATA: {
        code: 4013,
        message: 'Cannot find route data for "{0}"',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
    MISSING_SET_ID: {
        code: 4014,
        message: 'Routing specifier is invalid: "{0}"',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
    INVALID_ROUTE_DATA: {
        code: 4015,
        message: 'Route data validation failed: Route with id "{0}" from {1}.json must have a valid "{2}" property',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
    INVALID_NAMESPACE: {
        code: 4017,
        message: 'Route data validation failed: Route with id "{0}" from {1}.json must not have "-" in its "view" namespace',
        level: errors_1.DiagnosticLevel.Fatal,
        url: '',
    },
};
