// @webruntime/api/dist/commonjs/services/request-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RequestOutputTypes;
(function (RequestOutputTypes) {
    RequestOutputTypes["JSON"] = "json";
    RequestOutputTypes["COMPONENT"] = "component";
})(RequestOutputTypes = exports.RequestOutputTypes || (exports.RequestOutputTypes = {}));
function supportsRequests(service) {
    return (service.request !== undefined &&
        typeof service.request === 'function');
}
exports.supportsRequests = supportsRequests;
function containsComponentResource(output) {
    const compOutput = output;
    if (compOutput.resource) {
        return (compOutput.resource.code !== undefined && typeof compOutput.resource.code === 'string');
    }
    return false;
}
exports.containsComponentResource = containsComponentResource;
function containsJsonResource(output) {
    const jsonOutput = output;
    if (jsonOutput.resource) {
        return (jsonOutput.resource.json !== undefined && typeof jsonOutput.resource.json === 'object');
    }
    return false;
}
exports.containsJsonResource = containsJsonResource;
