// @communities-webruntime/design/src/modules/webruntimedesign/designmode/shared/constants.js
/**
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

export const BASE_Z_INDEX = 99999;

export const POLLING_INTERVAL = 100;

export const ELEMENT_ID_ATTRIBUTE_NAME = 'data-item-id';

export const ELEMENT_FQN_ATTRIBUTE_NAME = 'data-fqn';

export const COMPONENT_LABEL_ATTRIBUTE_NAME = 'data-component-label';

export const ELEMENT_WRAPPER_CLASS = 'interactions-element';

export const COMPONENT_WRAPPER_CLASS = 'interactions-component';

export const REGION_LABEL_ATTRIBUTE_NAME = 'data-region-label';

export const REGION_NAME_ATTRIBUTE_NAME = 'data-region-name';

export const REGION_WRAPPER_CLASS = 'interactions-region';

export const REGION_WRAPPER_MANUAL_CLASS = 'manual';

export const REGION_WRAPPER_DROPZONE_CLASS = 'dropzone';

export const COMPONENT_WRAPPER_ELEMENT_NAME = 'webruntimedesign-component-wrapper';

export const SECTION_FQN = 'community_layout/section';

export const PROXY_LABEL_CLASS = 'proxy-select-label';

export const CONTENT_REGION_NAME = 'content';
