// @webruntime/api/dist/commonjs/services/addressable-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const path_to_regexp_1 = require("path-to-regexp");
/**
 * The Base Addressable Service API Class
 *
 * Create a class extended from the AddressableService base class and implementing one of the available Service interfaces.
 * ```
 * import { AddressableService } from "@webruntime/api";
 * export class MyService extends AddressableService {
 * }
 * ```
 */
class AddressableService {
    constructor(uri) {
        this._uri = uri;
    }
    get uri() {
        return this._uri;
    }
    parseUrl(url) {
        const uris = Array.isArray(this.uri) ? this.uri : [this.uri];
        const params = uris
            .map(uri => {
            const keys = [];
            const pattern = path_to_regexp_1.pathToRegexp(`(.*)${uri}`, keys);
            const parts = pattern.exec(url);
            if (!parts) {
                return null;
            }
            return keys.reduce((obj, key, i) => {
                // Only keep "named parameters" if key.name === 0, this is our resourceRoot
                if (key.name !== 0) {
                    obj[key.name] = parts[i + 1];
                }
                return obj;
            }, {});
        })
            .filter(parts => parts !== null)[0];
        return params;
    }
    shutdown() { }
}
exports.AddressableService = AddressableService;
