// @webruntime/services/src/messages.ts
/** @hidden */
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { DiagnosticLevel } from '@lwc/errors';

export const bootstrapMessages = {
    MISSING_APP_CONFIG: {
        code: 7000,
        message: 'Application configuration is required to bootstrap',
        level: DiagnosticLevel.Error,
        url: '',
    },
};

export const importMapMessages = {
    MISSING_NAMES: {
        code: 7001,
        message: 'Cannot find a name or namespace from "{0}"',
        level: DiagnosticLevel.Error,
        url: '',
    },
    INVALID_PACKAGE: {
        code: 7002,
        message: 'No packaged URL found for packaged specifier "{0}"',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MISSING_PARAMS: {
        code: 7003,
        message: 'The "mode" and "locale" parameters are both required to create an import-map.',
        level: DiagnosticLevel.Error,
        url: '',
    },
};
