// @webruntime/server/dist/commonjs/container/resource-bundle.js
"use strict";
/** @hidden */
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const stream_1 = require("stream");
const api_1 = require("@webruntime/api");
const messages_1 = require("../utils/messages");
const { error } = console;
class ResourceBundle {
    constructor({ uid, path, specifier, results }) {
        this.uid = uid;
        this.path = path;
        this.specifier = specifier;
        this.results = Array.isArray(results) ? results : [results];
    }
    getReadableStream() {
        const stream = new stream_1.Readable();
        // push each resources code to the stream.
        this.results.forEach(result => {
            if (api_1.containsComponentResource(result)) {
                if (result.resource) {
                    stream.push(result.resource.code);
                }
            }
            if (api_1.containsJsonResource(result)) {
                if (result.resource) {
                    try {
                        stream.push(JSON.stringify(result.resource.json));
                    }
                    catch (e) {
                        error(api_1.generateMessage(messages_1.containerMessages.INVALID_DATA, [this.specifier, e]));
                    }
                }
            }
        });
        stream.push(null);
        return stream;
    }
}
exports.ResourceBundle = ResourceBundle;
