// @webruntime/navigation/src/modules/webruntime_navigation/utils/messages.js
/** @hidden */
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

const DiagnosticLevel = {
    Fatal: 0,
    Error: 1,
    Warning: 2,
    Log: 3,
};

function replaceArgs(message, args) {
    return Array.isArray(args)
        ? message.replace(/\{([0-9]+)\}/g, (_, index) => {
              return args[index];
          })
        : message;
}

export function generateMessage(info, args) {
    return `LWR${info.code}: ${replaceArgs(info.message, args)}`;
}

export function generateMessageObject(info, args) {
    return {
        ...info,
        code: `LWR${info.code}`,
        message: replaceArgs(info.message, args),
    };
}

export function invariant(condition, errorInfo, args) {
    if (!condition) {
        throw new Error(generateMessage(errorInfo, args));
    }
}

export const messages = {
    INVALID_MIXIN_CMP: {
        code: 4001,
        message: '{0} must be an Element type',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MISSING_CONTEXT: {
        code: 4002,
        message: 'Could not find context to perform navigation action.',
        level: DiagnosticLevel.Error,
        url: '',
    },
    INVALID_CONTEXT: {
        code: 4003,
        message:
            'Cannot register navigation context; it must have this shape: { navigate, generateUrl, subscribe }',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MULTIPLE_ROOTS: {
        code: 4004,
        message: 'Router connection failed. There can only be one root router.',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MULTIPLE_CHILDREN: {
        code: 4005,
        message:
            'Could not add to the navigation hierarchy. There can only be one child per navigation node.',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MISSING_ROUTE: {
        code: 4006,
        message: 'A route cannot be created to navigate to URL "{0}"',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MISSING_URL: {
        code: 4007,
        message: 'A URL cannot be created to navigate to route "{0}"',
        level: DiagnosticLevel.Error,
        url: '',
    },
    PRENAV_FAILED: {
        code: 4008,
        message: 'A preNavigate hook listener blocked routing to "{0}"',
        level: DiagnosticLevel.Warning,
        url: '',
    },
    MISSING_PATH: {
        code: 4009,
        message: 'A route definition must contain a "path" property.',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MISSING_ROUTE_CMP: {
        code: 4016,
        message: 'Expected a route view component with a default export.',
        level: DiagnosticLevel.Error,
        url: '',
    },
    MISSING_DATA_CONTEXT: {
        code: 4018,
        message: 'Could not find context to retrieve navigation data.',
        level: DiagnosticLevel.Error,
        url: '',
    },
};
