// @webruntime/services/dist/commonjs/messages.js
"use strict";
/** @hidden */
/**
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@lwc/errors");
exports.bootstrapMessages = {
    MISSING_APP_CONFIG: {
        code: 7000,
        message: 'Application configuration is required to bootstrap',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
};
exports.importMapMessages = {
    MISSING_NAMES: {
        code: 7001,
        message: 'Cannot find a name or namespace from "{0}"',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
    INVALID_PACKAGE: {
        code: 7002,
        message: 'No packaged URL found for packaged specifier "{0}"',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
    MISSING_PARAMS: {
        code: 7003,
        message: 'The "mode" and "locale" parameters are both required to create an import-map.',
        level: errors_1.DiagnosticLevel.Error,
        url: '',
    },
};
